File: client_sdk/python/mail_client.py
import hmac
import hashlib
import base64
import time
import requests
import os

class SecureMailClient:
    def __init__(self, base_url, api_key, secret_key, user_id):
        self.base_url = base_url
        self.api_key = api_key
        self.secret_key = secret_key
        self.user_id = user_id

    def _generate_signature(self, timestamp):
        # Format: apiKey + userId + timestamp
        data_to_sign = f"{self.api_key}{self.user_id}{timestamp}"
        
        # HMAC-SHA256
        signature = base64.b64encode(
            hmac.new(
                self.secret_key.encode('utf-8'),
                data_to_sign.encode('utf-8'),
                hashlib.sha256
            ).digest()
        ).decode('utf-8')
        return signature

    def send_mail(self, to, subject, body, from_addr, attachments=None):
        timestamp = str(int(time.time() * 1000))
        signature = self._generate_signature(timestamp)

        headers = {
            "X-API-KEY": self.api_key,
            "X-USER-ID": self.user_id,
            "X-SIGNATURE": signature,
            "X-TIMESTAMP": timestamp
        }

        payload = {
            "from": from_addr,
            "to": to,
            "subject": subject,
            "body": body
        }

        files = []
        if attachments:
            for path in attachments:
                if os.path.exists(path):
                    files.append(('attachments', open(path, 'rb')))

        try:
            url = f"{self.base_url}/mail"
            print(f"Sending request to {url}...")
            response = requests.post(url, headers=headers, data=payload, files=files)
            
            # Close files
            for _, f in files:
                f.close()

            print(f"Response Status: {response.status_code}")
            print(f"Response Body: {response.text}")
            return response.json()
        except Exception as e:
            print(f"Error: {e}")
            return None

# Usage Example
if __name__ == "__main__":
    # Replace with your provided credentials
    CLIENT_API_KEY = "dlt-dev-api-key-123"
    CLIENT_SECRET_KEY = "dlt-dev-secret-key-xyz"
    CLIENT_USER_ID = "employee_test_01"
    
    client = SecureMailClient(
        base_url="http://localhost:8080/api/v1",
        api_key=CLIENT_API_KEY, 
        secret_key=CLIENT_SECRET_KEY, 
        user_id=CLIENT_USER_ID
    )

    # Send an email
    client.send_mail(
        to="recipient@example.com",
        subject="Hello from Python SDK",
        body="<h1>This is a secure test</h1>",
        from_addr="dev-support@deloitte.com"
        # attachments=["document.pdf"]
    )

File: client_sdk/nodejs/mail_client.js
const axios = require('axios');
const crypto = require('crypto');
const FormData = require('form-data');
const fs = require('fs');

class SecureMailClient {
    constructor(baseUrl, apiKey, secretKey, userId) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.secretKey = secretKey;
        this.userId = userId;
    }

    generateSignature(timestamp) {
        // Format: apiKey + userId + timestamp
        const dataToSign = `${this.apiKey}${this.userId}${timestamp}`;

        // HMAC-SHA256
        return crypto.createHmac('sha256', this.secretKey)
            .update(dataToSign)
            .digest('base64');
    }

    async sendMail({ to, subject, body, from, attachments = [] }) {
        const timestamp = Date.now().toString();
        const signature = this.generateSignature(timestamp);

        const form = new FormData();
        form.append('from', from);
        form.append('to', to);
        form.append('subject', subject);
        form.append('body', body);

        attachments.forEach(filePath => {
            if (fs.existsSync(filePath)) {
                form.append('attachments', fs.createReadStream(filePath));
            }
        });

        const headers = {
            'X-API-KEY': this.apiKey,
            'X-USER-ID': this.userId,
            'X-TIMESTAMP': timestamp,
            'X-SIGNATURE': signature,
            ...form.getHeaders()
        };

        try {
            const url = `${this.baseUrl}/mail`;
            console.log(`Sending request to ${url}...`);

            const response = await axios.post(url, form, { headers });

            console.log('Response Status:', response.status);
            console.log('Response Body:', response.data);
            return response.data;
        } catch (error) {
            if (error.response) {
                console.error('Error Status:', error.response.status);
                console.error('Error Data:', error.response.data);
            } else {
                console.error('Error:', error.message);
            }
        }
    }
}

// Usage Example
(async () => {
    // Replace with your provided credentials
    const client = new SecureMailClient(
        'http://localhost:8080/api/v1',
        'dlt-dev-api-key-123',
        'dlt-dev-secret-key-xyz',
        'employee_test_01'
    );

    await client.sendMail({
        to: 'recipient@example.com',
        from: 'dev-support@deloitte.com',
        subject: 'Hello from Node.js SDK',
        body: '<h1>This is a secure test</h1>'
        // attachments: ['document.pdf']
    });
})();
